pkg/pillar/cmd/zedagent/attesttask.go

```diff
@@ -24,19 +32,197 @@ type WatchdogImpl struct{}
+	//Nonce for the current attestation cycle
+	Nonce []byte
+	//Quote for the current attestation cycle
+	InternalQuote *types.AttestQuote
+	//Iteration keeps track of retry count
+	Iteration int
+}
+
+const (
+	watchdogInterval  = 15
+	retryTimeInterval = 15
+)
+
+//One shot send, if fails, return an error to the state machine to retry later
+func trySendToController(attestReq *attest.ZAttestReq, iteration int) (*http.Response, []byte, types.SenderResult, error) {
+	data, err := proto.Marshal(attestReq)
+	if err != nil {
+		log.Fatal("SendInfoProtobufStr proto marshaling error: ", err)
+	}
+
+	buf := bytes.NewBuffer(data)
+	size := int64(proto.Size(attestReq))
+	attestURL := zedcloud.URLPathString(serverNameAndPort, zedcloudCtx.V2API,
+		devUUID, "attest")
+	return zedcloud.SendOnAllIntf(&zedcloudCtx, attestURL,
+		size, buf, iteration, true)
 }
 
 //SendNonceRequest implements SendNonceRequest method of zattest.Verifier
 func (server *VerifierImpl) SendNonceRequest(ctx *zattest.Context) error {
-	//XXX: Fill it in when Controller code is ready
+	if ctx.OpaqueCtx == nil {
+		log.Fatalf("[ATTEST] Uninitialized access to OpaqueCtx")
+	}
+	attestCtx, ok := ctx.OpaqueCtx.(*attestContext)
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected type from opaque ctx: %T",
+			ctx.OpaqueCtx)
+	}
+	if len(attestCtx.Nonce) > 0 {
+		//Clear existing nonce before attempting another nonce request
+		unpublishAttestNonce(attestCtx)
+		attestCtx.Nonce = nil
+	}
+	var attestReq = &attest.ZAttestReq{}
+
+	// bail if V2API is not supported
+	if !zedcloud.UseV2API() {
+		return zattest.ErrControllerReqFailed
+	}
+
+	attestReq.ReqType = attest.ZAttestReqType_ATTEST_REQ_NONCE
+	log.Debugf("Sending Nonce request %v", attestReq)
+	_, contents, senderStatus, err := trySendToController(attestReq, attestCtx.Iteration)
+	if err != nil || senderStatus != types.SenderStatusNone {
+		log.Errorf("[ATTEST] Error %v, senderStatus %v",
+			err, senderStatus)
+		//Increment Iteration for interface rotation next time
+		attestCtx.Iteration++
+		return zattest.ErrControllerReqFailed
+	}
+
+	//Reset Iteration on successful reply, irrespective of the reply content
+	attestCtx.Iteration = 0
+
+	attestResp := &attest.ZAttestResponse{}
+	if err := proto.Unmarshal(contents, attestResp); err != nil {
+		log.Errorf("[ATTEST] Error %v in Unmarshaling nonce response", err)
+		return zattest.ErrControllerReqFailed
+	}
+
+	respType := attestResp.GetRespType()
+	if respType != attest.ZAttestRespType_ATTEST_RESP_NONCE {
+		log.Errorf("[ATTEST] Got %v, but want %v",
+			respType, attest.ZAttestRespType_ATTEST_RESP_NONCE)
+		return zattest.ErrControllerReqFailed
+	}
+
+	if nonceResp := attestResp.GetNonce(); nonceResp == nil {
+		log.Errorf("[ATTEST] Got empty nonce response")
+		return zattest.ErrControllerReqFailed
+	} else {
+		attestCtx.Nonce = nonceResp.GetNonce()
+	}
+
+	return nil
+}
+
+func getPCRValues(internalQuote *types.AttestQuote, quoteMsg *attest.ZAttestQuote) error {
```

Comment:
"get" is odd as a name. This encodes the AttestQuote as ZAttestQoute, right?

Reply:
changed it to "encode..." instead of "get..."